src/main/java/io/cdap/plugin/servicenow/source/ServiceNowBaseSourceConfig.java

```diff
@@ -0,0 +1,341 @@
+    IdUtils.validateReferenceName(referenceName, collector);
+
+    validateCredentials(collector);
+    validateValueType(collector);
+    validateDateRange(collector);
+    validateTableNameField(collector);
+  }
+
+  public void validateCredentials(FailureCollector collector) {
+    if (!shouldConnect()) {
+      return;
+    }
+
+    if (Util.isNullOrEmpty(clientId)) {
+      collector.addFailure("Client ID must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_ID);
+    }
+
+    if (Util.isNullOrEmpty(clientSecret)) {
+      collector.addFailure("Client Secret must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_SECRET);
+    }
+
+    if (Util.isNullOrEmpty(restApiEndpoint)) {
+      collector.addFailure("API Endpoint must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_API_ENDPOINT);
+    }
+
+    if (Util.isNullOrEmpty(user)) {
+      collector.addFailure("User name must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_USER);
+    }
+
+    if (Util.isNullOrEmpty(password)) {
+      collector.addFailure("Password must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_PASSWORD);
+    }
+
+    validateServiceNowConnection(collector);
+  }
+
+
+  @VisibleForTesting
+  void validateServiceNowConnection(FailureCollector collector) {
+    try {
+      ServiceNowTableAPIClientImpl restApi = new ServiceNowTableAPIClientImpl(this);
+      restApi.getAccessToken();
+    } catch (Exception e) {
+      collector.addFailure("Unable to connect to ServiceNow Instance.",
+          "Ensure properties like Client ID, Client Secret, API Endpoint, User Name, Password " +
+            "are correct.")
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_ID)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_SECRET)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_API_ENDPOINT)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_USER)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_PASSWORD)
+        .withStacktrace(e.getStackTrace());
+    }
+  }
+
+  /**
+   * Returns the value type chosen.
+   *
+   * @param collector The failure collector to collect the errors
+   * @return An instance of SourceValueType
+   */
+  public SourceValueType getValueType(FailureCollector collector) {
+    SourceValueType type = getValueType();
+    if (type != null) {
+      return type;
+    }
+
+    collector.addFailure("Unsupported type value: " + valueType,
+        String.format("Supported value types are: %s", SourceValueType.getSupportedValueTypes()))
+      .withConfigProperty(ServiceNowConstants.PROPERTY_VALUE_TYPE);
+    collector.getOrThrowException();
+    return null;
+  }
+
+  /**
+   * Returns the value type chosen.
+   *
+   * @return An instance of SourceValueType
+   */
+  @Nullable
+  public SourceValueType getValueType() {
```

Comment:
This method is being referenced from several places within the code.

1. `ServiceNowMultiSourceConfig.validateTableNames(FailureCollector collector)` (Different class, Same package)
2. `ServiceNowSourceConfig.validateTableQueryMode(FailureCollector collector)` (Different class, Same package)
3. `ServiceNowTableAPIClientImpl.fetchTableRecords(String tableName, String startDate, String endDate, int offset,
                                                     int limit)` (Different class, Different package)
4. `ServiceNowTableAPIClientImpl.fetchTableSchemaUsingFirstRecord(String tableName, String startDate,
                                                                       String endDate, boolean fetchRecordCount)` (Different class, Different package)

I can pass in the `collector` argument in the first two usages and make the call to `SourceValueType getValueType(FailureCollector collector)` (the one having formal parameter) instead of `public SourceValueType getValueType()` but for the last two usages, I do not have the `collector` object with me to pass as an argument. I'll have to pass `null` or create a new `FailureCollector` object and then pass it as an actual parameter.

Let me know what you think.